### PR TITLE
Kotlin module generator

### PR DESCRIPTION
**Pull request by wesbillman**

### CODE REVIEW

File: kotlin-runtime/ftl-plugin/src/main/kotlin/xyz/block/ftl/gradle/SchemaGenerator.kt [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        )
+      }
+
+      klass.addFunction(
+        funSpecBuilder.build()
+      )
+    }
+
+    file.addType(klass.build())
+    file.build().writeTo(System.out)
```
Comment (wesbillman) [inline kotlin-runtime/ftl-plugin/src/main/kotlin/xyz/block/ftl/gradle/SchemaGenerator.kt:143]: Currently only writing to `System.out` but I should be able to send this to a directory as well. this was just faster for iterating on changes.

File: kotlin-runtime/ftl-plugin/src/main/kotlin/xyz/block/ftl/gradle/SchemaGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+          .build()
+      )
+      .baseUrl(ftlEndpoint)
+      .build()
+  }
+
+  fun generate() {
+    runBlocking {
+      launch {
+        grpcClient.create(ControllerServiceClient::class)
```
Comment (alecthomas) [inline kotlin-runtime/ftl-plugin/src/main/kotlin/xyz/block/ftl/gradle/SchemaGenerator.kt:39]: I'd probably split this into two classes: one that pulls the schema from the server and an `internal` one that just accepts a proto-encoded schema and writes Kotlin code.

File: kotlin-runtime/ftl-plugin/src/main/kotlin/xyz/block/ftl/gradle/SchemaGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+          .callTimeout(Duration.ofSeconds(10))
+          .protocols(listOf(Protocol.H2_PRIOR_KNOWLEDGE))
+          .build()
+      )
+      .baseUrl(ftlEndpoint)
+      .build()
+  }
+
+  fun generate() {
+    runBlocking {
```
Comment (alecthomas) [inline kotlin-runtime/ftl-plugin/src/main/kotlin/xyz/block/ftl/gradle/SchemaGenerator.kt:37]: Are we generating async clients? Should/can we generate threading clients?

File: kotlin-runtime/ftl-plugin/src/main/kotlin/xyz/block/ftl/gradle/SchemaGenerator.kt [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        funSpecBuilder.build()
+      )
+    }
+
+    file.addType(klass.build())
+    file.build().writeTo(System.out)
+  }
+
+  // TODO(wes): Is there a better way to get the type from this schema?
+  private fun getTypeName(typeInstance: Type): String {
```
Comment (alecthomas) [inline kotlin-runtime/ftl-plugin/src/main/kotlin/xyz/block/ftl/gradle/SchemaGenerator.kt:147]: Can't you just call .toString() on the base type? IIUC every Kotlin type has that method?

**Review by alecthomas — APPROVED** [review]

File: kotlin-runtime/ftl-plugin/src/main/kotlin/xyz/block/ftl/gradle/SchemaGenerator.kt (hunk 3 above)
Comment (wesbillman) [reply kotlin-runtime/ftl-plugin/src/main/kotlin/xyz/block/ftl/gradle/SchemaGenerator.kt:37]: Not sure I follow this one.

File: kotlin-runtime/ftl-plugin/src/main/kotlin/xyz/block/ftl/gradle/SchemaGenerator.kt (hunk 4 above)
Comment (wesbillman) [reply kotlin-runtime/ftl-plugin/src/main/kotlin/xyz/block/ftl/gradle/SchemaGenerator.kt:147]: Yeah I was def trying to find a better way to do this. From the Wire generated code, I don't see an obvious way to get the type, but I'm _very_ open to suggestions here. I can keep playing with it as well.

**Pull request merged**